x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,440 @@
+package keeper
+
+import (
+	errorsmod "cosmossdk.io/errors"
+	storetypes "cosmossdk.io/store/types"
+	"encoding/binary"
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/interchain-security/v5/x/ccv/provider/types"
+	"time"
+)
+
+// ConsumerPhase captures the phases of a consumer chain according to `docs/docs/adrs/adr-018-permissionless-ics.md`
+type ConsumerPhase byte
+
+const (
+	Registered ConsumerPhase = iota
```

Comment:
Would recommend some description in the code itself here, too, so that it shows up when you hover over the individual consts